### PR TITLE
Implement adjoint autodiff for `mesolve`

### PR DESCRIPTION
**Pull request by gautierronan**
Copy of https://github.com/PierreGuilmin/torchqdynamics/pull/39 after folder reorganization.

---

When calling `odeint(..., gradient_alg='adjoint',...)`, the following calls are sequentially made:
 - **Forward pass**: `odeint` --> `_odeint_adjoint` --> `ODEIntAdjoint.forward` --> `_odeint_inplace` --> `_odeint_main` --> `_fixed_odeint`/`_adaptive_odeint` --> `qsolver.forward`
 - **Backward pass**: `odeint` --> `_odeint_adjoint` --> `ODEIntAdjoint.backward` --> `_odeint_augmented_main` --> `_fixed_odeint_augmented`/`_adaptive_odeint_augmented` --> `qsolver.backward_augmented`

### CODE REVIEW

**Review by abocquet — APPROVED** [review]

File: torchqdynamics/odeint.py [hunk 1]
```diff
@@ -157,6 +165,156 @@ def _fixed_odeint(
...
+        between every checkpoint of the forward pass, as defined by `t_save`. This
+        helps with the stability of the backward integration.
+        Throughout this function, `y` is the state, `a = dL/dy` is the adjoint state,
+        and `g = dL/dp` is the gradient w.r.t. the parameters, where `L` is the loss
+        function and `p` the parameters.
+        TODO: Check that stacking `y` and `a` does not improve performance of the
+        solver. There was funky stuff happening in this regard in some tests I ran.
+        """
+        # disable gradient computation
+        torch.set_grad_enabled(False)
```
Comment (PierreGuilmin) [inline torchqdynamics/odeint.py:276]: Should we use this as a context manager rather than just disabling it?

**Review by PierreGuilmin — CHANGES_REQUESTED** [review]
🚀 Very minor comments, this is really well written, congrats! It was much easier to review than I expected. I'd like to read it again tmo or Wednesday once you have fixed the comments, because I just read quickly the core of the PR, but it looks all good.

A few remarks:
- I didn't review the order 2, I trust you for it (I keep it in a todo list for the day I understand properly the adjoint).
- What about order 1.5? (I'm 100% fine to do it in another PR, just curious if we know how to do it or not)
- Can you do some test on your own just to make sure everything runs through? (no need to check that the computation is correct, neither to add a test file, just a quick check to remove dumb bugs)
- How much of the code is inspired from torchdiffeq/torchdyn? We might want to check their license if it is close.
- Could you add the PR description as a comment on top of the `_odeint_adjoint` function? It really helps understanding how all of this is tied tgther.
- I think it would be amazing to have, once our doc website is up and running, a short documentation page to explain what is the adjoint method, what are the advantages, how it works and how we implemented it 😉 Just saying so you can gather ideas / examples in your mind for the day we write this page.

File: torchqdynamics/mesolve/rouchon.py [hunk 2]
```diff
@@ -19,8 +21,8 @@ def __init__(self, H: TDOperator, jump_ops: Tensor, solver_options: SolverOption
         self.H = H[:, None, ...]  # (b_H, 1, n, n)
         self.jump_ops = jump_ops[None, ...]  # (1, len(jump_ops), n, n)
         self.sum_nojump = (jump_ops.adjoint() @ jump_ops).sum(dim=0)  # (n, n)
-        n = H.shape[-1]
-        self.I = torch.eye(n).to(H)  # (n, n)
+        self.n = H.shape[-1]
```
[the remark this answers is not in the document]
Comment (gautierronan) [reply torchqdynamics/mesolve/rouchon.py:24]: Not for now but ça mange pas de pain

File: torchqdynamics/solver_utils.py [hunk 3]
```diff
@@ -45,3 +47,16 @@ def bexpect(operators: Tensor, state: Tensor) -> Tensor:
     # TODO: Once QTensor is implemented, check if state is a density matrix or ket.
     # For now, we assume it is a density matrix.
     return torch.einsum('bij,...ji->...b', operators, state)
+
+
+def none_to_zeros_like(input, shaping_tuple):
```
[the remark this answers is not in the document]
Comment (gautierronan) [reply torchqdynamics/solver_utils.py:52]: Yes this is done in the following PR.

**Comment by gautierronan** [comment]
> 🚀 Very minor comments, this is really well written, congrats! It was much easier to review than I expected. I'd like to read it again tmo or Wednesday once you have fixed the comments, because I just read quickly the core of the PR, but it looks all good.

Thanks for the nice comments :)

> A few remarks:
> 
>     * I didn't review the order 2, I trust you for it (I keep it in a todo list for the day I understand properly the adjoint).

It should be quite similar to order 1. It's just a matter of solving rho backwards, and solving phi in Heisenberg forward essentially.

>     * What about order 1.5? (I'm 100% fine to do it in another PR, just curious if we know how to do it or not)

It should work similarly, I was just lazy to do it and didn't want to overload the PR.

>     * Can you do some test on your own just to make sure everything runs through? (no need to check that the computation is correct, neither to add a test file, just a quick check to remove dumb bugs)

I will do that at some point for sure. But I actually don't have any torchqdynamics code to test it on. However, the overall gradient computation should work this I have a similar implementation for my project that works. But there are for sure dumb bugs remaining here. 

>     * How much of the code is inspired from torchdiffeq/torchdyn? We might want to check their license if it is close.

To be fair, not a lot at this point. I really moved away from their implementations, so I think we don't need anything in this regard.

>     * I think it would be amazing to have, once our doc website is up and running, a short documentation page to explain what is the adjoint method, what are the advantages, how it works and how we implemented it 😉 Just saying so you can gather ideas / examples in your mind for the day we write this page.

Well we will publish a paper at some point to explain all of this. But yes, once its done I can adapt and make a simpler version for the docs !

File: torchqdynamics/odeint.py (hunk 1 above)
Comment (PierreGuilmin) [reply torchqdynamics/odeint.py:276]: Well here we deactivate the gradient calculation for the user for the rest of his session no? I don't think we should change global settings without consent.

I don't know the difference between `with torch.no_grad()` and `with torch.set_grad_enabled(True)`, you might want to read https://pytorch.org/docs/stable/notes/autograd.html#locally-disable-grad-doc to pick the correct one.

**Review by PierreGuilmin — APPROVED** [review]
Let's merge this, I think it's good to move forward, I'll read the code in details when I know more about the adjoint 😉

**Comment by gautierronan** [comment]
> Let's merge this, I think it's good to move forward, I'll read the code in details when I know more about the adjoint 😉

Alright I debugged the thing. Here's my simple debugging file. Might be useful to write a test at some point./

```python
import torch
import torch.nn as nn
import torchqdynamics as tq
import qutip as qt
from math import pi, sqrt

# parameters
d = 32
kappa = 1.5
delta = nn.Parameter(torch.tensor([4 * pi]))
alpha0 = 1.0 + 1.0j

# operators
a = tq.from_qutip(qt.destroy(d))
adag = a.adjoint()
H = delta * adag @ a
jump_ops = [sqrt(kappa) * a]
exp_ops = [(a+adag)/sqrt(2), (a-adag)/(sqrt(2) * 1j)]

# parameters
parameters = (delta,)

# other arguments
rho0 = qt.coherent(d, alpha0)
tsave = torch.linspace(0.0, 1.0, 11)
solver = tq.solver.Rouchon1(dt=1e-3)

# solve
states, exp = tq.mesolve(H, jump_ops, rho0, tsave, exp_ops=exp_ops, solver=solver, gradient_alg='adjoint', parameters=parameters)

# backward pass
loss = states.sum()
loss.backward()

# result
print(states.shape)
print(exp.shape)
```

**Review by PierreGuilmin — APPROVED** [review]
